pkgs/build-support/build-fhs-userenv/chrootenv.c

```diff
@@ -20,56 +21,112 @@
 #include <sys/stat.h>
 #include <sys/wait.h>
 
-char *env_whitelist[] = {"TERM",
-                         "DISPLAY",
-                         "XAUTHORITY",
-                         "HOME",
-                         "XDG_RUNTIME_DIR",
-                         "LANG",
-                         "SSL_CERT_FILE",
-                         "DBUS_SESSION_BUS_ADDRESS"};
-
-char **env_build(char *names[], size_t len) {
-  char *env, **ret = malloc((len + 1) * sizeof(char *)), **ptr = ret;
-
-  for (size_t i = 0; i < len; i++) {
-    if ((env = getenv(names[i]))) {
-      if (asprintf(ptr++, "%s=%s", names[i], env) < 0)
-        errorf(EX_OSERR, "asprintf");
+#define LEN(x) sizeof(x) / sizeof(*x)
+
+char *env_blacklist[] = {};
+
+char **env_filter(char *envp[]) {
+  char **filtered_envp = malloc(sizeof(*envp));
+  size_t n = 0;
+
+  while (*envp != NULL) {
+    bool blacklisted = false;
+
+    for (size_t i = 0; i < LEN(env_blacklist); i++) {
+      if (!strncmp(*envp, env_blacklist[i], strlen(env_blacklist[i]))) {
+        blacklisted = true;
+        break;
+      }
+    }
+
+    if (!blacklisted) {
+      filtered_envp = realloc(filtered_envp, (n + 2) * sizeof(*envp));
```

Comment:
Leaks old memory location of `filtered_envp` in the error case -> this would require a temporary variable that is only copied to `filtered_envp` in the success case

Reply:
I'm not sure I understand. Below, `filtered_envp == NULL` handles the error case.

Reply:
`O(n^2)` :(

Reply:
By the time `filtered_envp` is NULL you have leaked the old `filtered_envp` array.

Reply:
Dropping this code as you suggested.

Reply:
> By the time filtered_envp is NULL you have leaked the old filtered_envp array.

I don't think I understand why this matters before an exit.